CMakeLists.txt

```diff
@@ -14,15 +14,26 @@ set(COMBINED_PROJECT TRUE)
 # Library type
 option(STATIC "Set to ON to build xlnt as a static library instead of a shared library" OFF)
 
-# c++ language standard to use
-set(XLNT_VALID_LANGS 11 14 17)
-set(XLNT_CXX_LANG "14" CACHE STRING "c++ language features to compile with")
+# C++ language standard to use
+set(XLNT_CXX_VALID_LANGS 11 14 17)
+set(XLNT_CXX_LANG "14" CACHE STRING "C++ language features to compile with")
 # enumerate allowed values for cmake gui
-set_property(CACHE XLNT_CXX_LANG PROPERTY STRINGS ${XLNT_VALID_LANGS})
-# validate value is in XLNT_VALID_LANGS
-list(FIND XLNT_VALID_LANGS ${XLNT_CXX_LANG} index)
+set_property(CACHE XLNT_CXX_LANG PROPERTY STRINGS ${XLNT_CXX_VALID_LANGS})
+# validate value is in XLNT_CXX_VALID_LANGS
+list(FIND XLNT_CXX_VALID_LANGS ${XLNT_CXX_LANG} index)
 if(index EQUAL -1)
-    message(FATAL_ERROR "XLNT_CXX_LANG must be one of ${XLNT_VALID_LANGS}")
+    message(FATAL_ERROR "XLNT_CXX_LANG must be one of ${XLNT_CXX_VALID_LANGS}")
+endif()
+
+# C language standard to use
+set(XLNT_C_VALID_LANGS 11 17 23)
+set(XLNT_C_LANG "11" CACHE STRING "C language features to compile with")
```

Comment:
Is it a good idea to set the C standard independently of the C++ standard? If the user configures the C++ version, wouldn't he expect that the C version is also changed to the "corresponding" C version?

Reply:
I was not sure about this myself and added the option to change the C version in order to test the behavior. I tested it with GCC 14.2.1 on Fedora 41, and I can say that GCC ignores `__STDC_VERSION__` when the code is compiled in C++ mode - but the behavior is implementation-defined, so it might be defined in other compilers.

However, after testing and searching for it, it seems that when the code is compiled using a C++ compiler, the C version changes automatically to whatever C version the C++ standard refers to. I think the compiler wouldn't be standard-conformant if it did otherwise.

This effectively means that the C version only affects code that is compiled as C (if you find something else, please let me know). How the different compilers define what should be compiled as C and what as C++ is, again, compiler-dependent. For example, by default, [MSVC compiles files](https://learn.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-170#c-standards-support-1) with a `.cpp` extension as C++, and files with a `.c` extension as C. [GCC behaves similarly](https://gcc.gnu.org/onlinedocs/gcc/Invoking-G_002b_002b.html), except that calling `g++` will always compile all code as C++, while `gcc` will treat `.c` files (important: lowercase `.c`) as C and the rest as C++.

In other words, you are right - it might not make a big difference in XLNT, but it does make a small difference: miniz, sha1, sha512 and some additional files in `libstudxml/details/genx` and `libstudxml/details/expat` have `.c` extensions, meaning that the code will be compiled as C on some compilers. What does this mean, effectively? If the C version is not defined and the code is compiled as C, I would expect the compiler to use the default version of C defined by the compiler. On MSVC, that is currently [C89 with a few C99 extensions](https://learn.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-170#c-standards-support-1), while on current versions of GCC, [it is `gnu17`](https://gcc.gnu.org/onlinedocs/gcc-12.1.0/gcc/Standards.html#C-Language).

I also want to mention that the only reason why I defined the C version is because I wanted to check whether our version checks work correctly and add them for `localtime_r` and `gmtime_r` when C23 is available - and then I left the code in, because I though it wouldn't be a bad idea, in the end.

Does this whole story make a huge difference? Probably not, as we don't have much C code. Explicitly defining the C version might be a bit nicer, as we have control over the version (although we could also do this internally, without giving the user the option to change it). However, in the end, I think we should make a decision but don't put too much effort into it, since it shouldn't make a huge difference.

Reply:
Maybe just mark the option as advanced.

Reply:
And setting to c99 by default? This corresponds with the default / minimal required c++ version. So, the user has not to change it for an old compiler.

Reply:
> Maybe just mark the option as advanced.

Done.

> And setting to c99 by default? This corresponds with the default / minimal required c++ version. So, the user has not to change it for an old compiler.

Keep in mind that XLNT already uses C++14 by default, and not C++11.

I did some research and it seems that although C11 was standardized a few months after C++11, [GCC 4.9 contained full support for C11](https://cmake.org/cmake/help/latest/prop_tgt/C_STANDARD.html), while C++11 support was only [officially stable in GCC 5.1](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html) (but experimental beginning with GCC 4.8).

I also checked out the CMake documentation for the [`C_STANDARD` variable](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html) and they state two important things:

> For compilers that have no notion of a C standard level, such as Microsoft Visual C++ before VS 16.7, this property has no effect.

> with a compiler which does not support -std=gnu11 or an equivalent flag will not result in an error or warning, but will instead add the -std=gnu99 or -std=gnu90 flag if supported. This "decay" behavior may be controlled with the [C_STANDARD_REQUIRED](https://cmake.org/cmake/help/latest/prop_tgt/C_STANDARD_REQUIRED.html#prop_tgt:C_STANDARD_REQUIRED) target property.

The [`CXX_STANDARD` variable](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html) behaves very similarly. Due to the fact that C11 compilers are already very old, and old versions of MSVC simply won't use the C version, I think we are on the safe side already.

Otherwise we could also think about downgrading to C++11 and C99 by default (instead of C++14 and C11), but is it worth it? I'm not sure it's really worth it, as compilers supporting C++14 and C11 are already available since 10 years or more.